train.py

```diff
@@ -293,14 +335,33 @@ def loss_fn(pred, labels):
 
             input_ids = input_ids.cuda()
             labels = labels.cuda()
-
             optimizer.zero_grad()
 
-            # forward / backward
-            with loss_parallel_ctx():
-                pred = model(input_ids)
-                loss = loss_fn(pred, labels)
-                loss.backward()
+            if parallel_dims.pp_enabled:
+                # pipeline parallel forward / backward inside step() call
+                is_last_stage = pp_mesh.get_local_rank() == pp_mesh.size() - 1
+
+                with loss_parallel_ctx():
+                    if pp_mesh.get_local_rank() == 0:
+                        pp_schedule.step(input_ids)
+                    elif is_last_stage:
+                        losses = []
+                        pp_schedule.step(target=labels, losses=losses)
+                    else:
+                        schedule.step()
+
+                # accumulate losses across pipeline microbatches
+                loss = (
+                    torch.mean(torch.stack(losses))
+                    if is_last_stage
+                    else torch.Tensor([-1.0])
```

Comment:
Why we need the default -1 value? because of logging purpose?

Reply:
oh, yea i could make it a 'None' but then i have to update logger to not log at all. maybe that's actually a better way to do it. let me try that.

Reply:
ok- so what I could do is try to alter the metrics code so that on non-last-stage ranks, we omit printing loss, or, we print "loss: None" instead of -1.

The change will add more lines of code, since I need to deal with several places that expect loss and global_[avg/mean]_loss to be valid numbers
 - avoid writing them into metrics dict
 - replace their format string with a string value instead of a float value in the logger.info
 - avoid calling loss.item() in the first place

I agree in principle that's the "right" fix, but i'm not sure if its worth the LOC / complexity.  I don't totally hate the `-1` thing.

Another option I considered is to skip the whole codeblock of '# log metrics' on non-last-stage ranks.  I ruled this out, since it is still useful to log mfu, memory for other ranks.

So let me know what you want to do here.